src/coreclr/src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,871 @@
+
+    builder->Flush();
+
+    return builder->GetBlobLength() > 0;
+}
+
+MethodDesc* TailCallHelp::CreateStoreArgsStub(TailCallInfo& info)
+{
+    SigBuilder sigBuilder;
+    CreateStoreArgsStubSig(info, &sigBuilder);
+
+    DWORD cbSig;
+    PCCOR_SIGNATURE pSig = AllocateSignature(
+        info.LoaderAllocator, sigBuilder, &cbSig);
+
+    SigTypeContext emptyCtx;
+
+    ILStubLinker sl(info.Caller->GetModule(),
+                    Signature(pSig, cbSig),
+                    &emptyCtx,
+                    NULL,
+                    FALSE,
+                    FALSE);
+
+    ILCodeStream* pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+
+    DWORD bufferLcl = pCode->NewLocal(ELEMENT_TYPE_I);
+
+    void* pGcDesc = NULL;
+    if (info.HasGCDescriptor)
+    {
+        DWORD gcDescLen;
+        PVOID gcDesc = info.GCRefMapBuilder.GetBlob(&gcDescLen);
+        pGcDesc = AllocateBlob(info.LoaderAllocator, gcDesc, gcDescLen);
+    }
+
+    pCode->EmitLDC(info.ArgBufLayout.Size);
+    pCode->EmitLDC(DWORD_PTR(pGcDesc));
+    pCode->EmitCONV_I();
+    pCode->EmitCALL(METHOD__RUNTIME_HELPERS__ALLOC_TAILCALL_ARG_BUFFER, 2, 1);
+    pCode->EmitSTLOC(bufferLcl);
+
+    auto emitOffs = [&](UINT offs)
+    {
+        pCode->EmitLDLOC(bufferLcl);
+        if (offs != 0)
+        {
+            pCode->EmitLDC(offs);
+            pCode->EmitADD();
+        }
+    };
+
+    unsigned int argIndex = 0;
+
+    for (COUNT_T i = 0; i < info.ArgBufLayout.Values.GetCount(); i++)
+    {
+        const ArgBufferValue& arg = info.ArgBufLayout.Values[i];
+        CorElementType ty = arg.TyHnd.GetSignatureCorElementType();
+
+        emitOffs(arg.Offset);
+        pCode->EmitLDARG(argIndex++);
+        EmitStoreTyHnd(pCode, arg.TyHnd);
+    }
+
+    if (info.ArgBufLayout.HasTargetAddress)
+    {
+        emitOffs(info.ArgBufLayout.TargetAddressOffset);
+        pCode->EmitLDARG(argIndex++);
+        pCode->EmitSTIND_I();
+    }
+
+    pCode->EmitRET();
+
+    Module* pLoaderModule = info.Caller->GetLoaderModule();
+    MethodDesc* pStoreArgsMD =
+        ILStubCache::CreateAndLinkNewILStubMethodDesc(
+            info.LoaderAllocator,
+            pLoaderModule->GetILStubCache()->GetOrCreateStubMethodTable(pLoaderModule),
+            ILSTUB_TAILCALL_STOREARGS,
+            info.Caller->GetModule(),
+            pSig, cbSig,
+            &emptyCtx,
+            &sl);
+
+#ifdef _DEBUG
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: StoreArgs IL created\n"));
+    sl.LogILStub(CORJIT_FLAGS());
+#endif
+
+    return pStoreArgsMD;
+}
+
+void TailCallHelp::CreateStoreArgsStubSig(
+    const TailCallInfo& info, SigBuilder* sig)
+{
+    // The store-args stub will be different depending on the tailcall site.
+    // Specifically the following things might be conditionally inserted:
+    // * Call target address (for calli or generic calls resolved at tailcall site)
+    // * This pointer (for instance calls)
+    // * Generic context (for generic calls requiring context)
```

Comment:
It is still correct that this can store generic context? I do not see it in the method.

Reply:
No, this is left over from before we used instantiating stubs, I will remove this part of the comment.